src/arithmetic.jl

```diff
@@ -250,93 +248,85 @@ for T in (:HomogeneousPolynomial, :TaylorN)
     end
 end
 
+for (T, W) in ((:Taylor1, :Number), (:TaylorN, :NumberNotSeriesN))
+    @eval *{T<:$W, S<:$W}(a::$T{T}, b::$T{S}) = *(promote(a,b)...)
 
+    @eval function *{T<:$W}(a::$T{T}, b::$T{T})
+        corder = max(a.order, b.order)
+        if $T == Taylor1
+            c = Taylor1(zero(a[1]), corder)
+        else
+            c = TaylorN(zeros(HomogeneousPolynomial{T}, corder))
+        end
 
-doc"""
-```
-*(a, b)
-```
-
-Return the Taylor expansion of $a \cdot b$, of order `max(a.order,b.order)`, for
-`a::Taylor1`, `b::Taylor1` polynomials.
-
-For details on making the Taylor expansion, see [`TaylorSeries.mulHomogCoef`](@ref).
-"""
-*{T<:Number,S<:Number}(a::Taylor1{T}, b::Taylor1{S}) = *(promote(a,b)...)
-
-function *{T<:Number}(a::Taylor1{T}, b::Taylor1{T})
-    a, b = fixorder(a, b)
-    coeffs = similar(a.coeffs)
-    @inbounds for k = 0:a.order
-        coeffs[k+1] = mulHomogCoef(k, a.coeffs, b.coeffs)
+        @inbounds for ord = 0:corder
+            mul!(c, a, b, ord) # updates c[ord+1]
```

Comment:
Let me try to explain the "design principle". Essentially, we need the same mutating functions for the coefficients of either `Taylor1` or `TaylorN`, which must include "what to update" (`c`), the info to update it (`a` and `b` here because this is a binary function), and the order of the coefficient to be updated. So, as far as I can see, this is some sort of minimum required. For `mul!` I tried to be as close as what we already had for `HomogeneousPolynomial`.

Do you have some suggestion on this?